staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/values.go

```diff
@@ -0,0 +1,570 @@
+		}
+		return types.NewErr("invalid data, expected XIntOrString value to be either a string or integer")
+	}
+	if schema.Type == "object" {
+		m, ok := unstructured.(map[string]interface{})
+		if !ok {
+			types.NewErr("invalid data, expected map[string]interface{} to match the provided schema with type=object")
+		}
+		if schema.XEmbeddedResource || schema.Properties != nil {
+			return &unstructuredMap{
+				value:      m,
+				schema:     schema,
+				propSchema: func(key string) *structuralschema.Structural { schema := schema.Properties[key]; return &schema },
+			}
+		}
+		if schema.AdditionalProperties != nil && schema.AdditionalProperties.Structural != nil {
+			return &unstructuredMap{
+				value:      m,
+				schema:     schema,
+				propSchema: func(key string) *structuralschema.Structural { return schema.AdditionalProperties.Structural },
+			}
+		}
+		types.NewErr("invalid object type, expected either Properties or AdditionalProperties with Allows=true and non-empty Schema")
+	}
+	if schema.Type == "array" {
+		if schema.Items == nil {
+			types.NewErr("invalid array type, expected Items with a non-empty Schema")
+		}
+		l, ok := unstructured.([]interface{})
+		if ok {
+			types.NewErr("invalid data, expected []interface{} to match the provided schema with type=array")
+		}
+		typedList := unstructuredList{elements: l, itemsSchema: schema.Items}
+		listType := schema.XListType
+		if listType != nil {
+			switch *listType {
+			case "map":
+				mapKeys := schema.Extensions.XListMapKeys
+				return &unstructuredMapList{unstructuredList: typedList, escapedKeyProps: escapeKeyProps(mapKeys)}
+			case "set":
+				return &unstructuredSetList{unstructuredList: typedList}
+			}
+		}
+		return &typedList
+	}
+
+	if str, ok := unstructured.(string); ok {
+		if schema.Type == "string" && schema.ValueValidation != nil {
+			switch schema.ValueValidation.Format {
+			case "duration":
+				d, err := strfmt.ParseDuration(str)
+				if err != nil {
+					return types.NewErr("Invalid duration %s: %v", str, err)
+				}
+				return types.DefaultTypeAdapter.NativeToValue(d)
+			case "date":
+				d, err := time.Parse(strfmt.RFC3339FullDate, str) // strfmt uses this format for OpenAPIv3 value validation
+				if err != nil {
+					return types.NewErr("Invalid date formatted string %s: %v", str, err)
+				}
+				return types.DefaultTypeAdapter.NativeToValue(d)
+			case "date-time":
+				d, err := strfmt.ParseDateTime(str)
+				if err != nil {
+					return types.NewErr("Invalid date-time formatted string %s: %v", str, err)
+				}
+				return types.DefaultTypeAdapter.NativeToValue(time.Time(d))
+			case "byte":
+				base64 := strfmt.Base64{}
+				err := base64.UnmarshalText([]byte(str))
+				if err != nil {
+					return types.NewErr("Invalid byte formatted string %s: %v", str, err)
+				}
+				return types.DefaultTypeAdapter.NativeToValue([]byte(base64))
+			}
+		}
+	}
+	return types.DefaultTypeAdapter.NativeToValue(unstructured)
```

Comment:
we're not comparing schema.Type to the data type of unstructured here like we did for object/map/array/int-or-string above... that means we can return a value that will disagree with the type checked expression, right?

Reply:
I'll match all cases and then return an error if none match.

Reply:
This teased out a bug here: https://github.com/kubernetes/kubernetes/pull/106051/commits/a56e86ffdc2e31d884933a3cb2e6f4d693a70a0d#diff-d2c88853433233f52c9c6a142c6da3bb5d038bd6f33a7d5818946ada51e7a42dR63

My test code was also getting overly lenient treatment so I fixed it after adding the strict schema checking.